packages/msal-interactor/src/MsalInteractor.ts

```diff
@@ -2,105 +2,230 @@
+	SilentRequest,
+} from '@azure/msal-browser'
+
+export interface MsalInteractorOptions {
+	msalConfig: Configuration
+	oidcScopes: string[]
+}
+
+export interface MsalInteractorLoginOptions {
+	usePopup?: boolean
+}
 
 export class MsalInteractor {
-	private _instance: UserAgentApplication | undefined
+	private static _instance: MsalInteractor | null
+	private static _options: MsalInteractorOptions | null
+	private _msalConfig: Configuration
+	private _oidcScopes: string[]
+	private _msalInstance: IPublicClientApplication
+	private _accountInfo: AccountInfo | null = null
+	public isAuthenticated: () => Promise<boolean>
 
-	public constructor(
-		private msalConfig: Configuration,
-		private scopes: string[],
-		private authDisabled = false,
-	) {}
+	/**
+	 * Singleton. Use MsalInteractor.configure(options)
+	 * and MsalInteractor.getInstance()
+	 */
+	private constructor({ msalConfig, oidcScopes }: MsalInteractorOptions) {
+		this._msalConfig = msalConfig
+		this._oidcScopes = oidcScopes
+		this._msalInstance = new PublicClientApplication(this._msalConfig)
+
+		// On page load/instantiation, check if responding to redirect auth flow
+		const pageLoadHandler = this._msalInstance
+			.handleRedirectPromise()
+			.then(this._handleAuthResponse)
+
+		this.isAuthenticated = () => pageLoadHandler.then(res => !!res)
+	}
+
+	public static configure = (options: MsalInteractorOptions): void => {
+		MsalInteractor._options = options
+	}
+
+	public static getInstance = (): MsalInteractor => {
+		if (MsalInteractor._instance) {
+			return MsalInteractor._instance
+		}
+
+		if (!MsalInteractor._options) {
+			throw new TypeError(
+				`Cannot instantiate MsalInteractor without configuration options. Call configure() before getInstance()`,
+			)
+		}
+
+		MsalInteractor._instance = new MsalInteractor(MsalInteractor._options)
+		return MsalInteractor._instance
+	}
 
 	/**
-	 * Gets the MSAL Instance
+	 * User login.
+	 *
+	 * Attempt to login silently with current valid session
+	 * before logging in using either popUp or redirect flow.
 	 */
-	public get instance(): UserAgentApplication {
-		if (this._instance != null) {
-			return this._instance
+	public login = async (
+		options: MsalInteractorLoginOptions = {},
+	): Promise<AuthenticationResult | null> => {
+		const { usePopup = false } = options
+
+		// try to login silently
+		if (await this.isAuthenticated()) {
+			const authResult = await this._getAuthResult()
+			if (authResult) {
+				return authResult
+			}
 		}
 
-		const result = new UserAgentApplication(this.msalConfig)
-		if (result != null) {
-			result.handleRedirectCallback((err, res) => {
-				console.log('MSAL handling redirect callback', err, res)
-			})
+		if (usePopup) {
+			const authResult = this._msalInstance
+				.loginPopup({
+					scopes: this._oidcScopes,
+				})
+				.then(this._handleAuthResponse)
+
+			this.isAuthenticated = () => authResult.then(res => !!res)
+			return authResult
+		} else {
+			return this._msalInstance
```

Comment:
Do we not need to set the isAuthenticated field in this branch?

Reply:
The new @azure/msal-browser package does not expose a simple mechanism for checking if someone is authenticated. One cannot simply check if there is a logged-in account on page load as populating the user accounts and the cache is async. Instead, one must register a listener with `hanldeRedirectPromise()` to check on page load if someone is authenticated. This works for the first registered handler. The problem is that if you register more than one `handleRedirectPromise()` handler then all subsequent handlers fire too soon, before the cache and accounts are populated. This results in false negatives, the library reporting nobody has authenticated when in fact someone is authenticated. I exposed this `.isAuthenticated` method to get around this issue and make it easier for end developers to check if someone is authenticated. This is the same issue that led me to design this package as a singleton.